src/net/socket.rs

```diff
@@ -112,71 +138,101 @@ pub struct Socket {
    rx: mpsc::UnboundedReceiver<relay::Packet>,
 }
 
+type Stream = SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>;
+type Sink = SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>;
+
 impl Socket {
    async fn read_packet(
-      read_half: &mut tcp::OwnedReadHalf,
-      len: usize,
+      data: Vec<u8>,
       output: &mut mpsc::UnboundedSender<relay::Packet>,
    ) -> anyhow::Result<()> {
-      if len > relay::MAX_PACKET_SIZE as usize {
+      if data.len() > relay::MAX_PACKET_SIZE as usize {
          anyhow::bail!("Packet is too big");
       }
-      let mut bytes = vec![0; len as usize];
-      read_half.read_exact(&mut bytes).await?;
-      let packet = bincode::deserialize(&bytes).context("Invalid packet")?;
+      let packet = bincode::deserialize(&data).context("Invalid packet")?;
       output.send(packet)?;
       Ok(())
    }
 
    async fn receiver_loop(
-      mut read_half: tcp::OwnedReadHalf,
+      mut stream: Stream,
       mut output: mpsc::UnboundedSender<relay::Packet>,
-      mut quit: oneshot::Receiver<Quit>,
+      quit: (broadcast::Sender<Quit>, broadcast::Receiver<Quit>),
    ) -> anyhow::Result<()> {
+      let (tx, mut rx) = quit;
       loop {
          tokio::select! {
             biased;
-            Ok(_) = &mut quit => {
+            Ok(_) = rx.recv() => {
                log::info!("receiver: received quit signal");
                break;
             },
-            len = read_half.read_u32() => Self::read_packet(
-               &mut read_half,
-               len? as usize,
-               &mut output
-            ).await?,
+            Some(message) = stream.next() => {
+               match message {
+                  Ok(Message::Binary(data)) => {
+                     Self::read_packet(data, &mut output).await?
+                  },
+                  Ok(Message::Close(frame)) => {
+                     bus::push(Fatal(anyhow::anyhow!("Relay has been disconnected.")));
+
+                     if let Some(frame) = frame {
+                        log::warn!("Relay has been disconnected, reason: \"{}\", code: {}", frame.reason, frame.code);
```

Comment:
Log messages should start with lowercase letters.
```suggestion
                        log::warn!("the relay has disconnected: {:?}, code: {}", frame.reason, frame.code);
```